fixes dhcp server virsh.rb to interpret prefix

thkrmr opened the pull request:
The virsh.rb loadSubnets method was only looking for a definition of a netmask inside the libvirt network XML, despite netmask not being there in case a prefix attribute was defined. netmask and prefix are mutually exclusive by libvirt definition.

Also the ipaddr library has no easy way of retrieving a readable netmask from a prefix, albeit this functionality is provided by the inspect method, in this case the same thing was done as in inspect.

thkrmr commented:
Previous behaviour resulted in an error which you can observe below. Note the XML, it's a valid libvirt network definition that works fine, but omits netmask in favor of prefix.

```
D, [2014-06-04T11:29:21.371624 #4056] DEBUG -- : Loading subnets for virsh_networksubnet
D, [2014-06-04T11:29:21.371748 #4056] DEBUG -- : /usr/bin/sudo /usr/bin/virsh net-dumpxml subnet 2>&1
E, [2014-06-04T11:29:21.392522 #4056] ERROR -- : DHCP virsh provider error: unable to retrive virsh info: undefined method `<' for nil:NilClass
D, [2014-06-04T11:29:21.392614 #4056] DEBUG -- : <network connections='1'>
  <name>subnet</name>
  <uuid>4543f7f0-9d75-6e18-833e-bc78854b733f</uuid>
  <forward dev='br0' mode='route'>
    <interface dev='br0'/>
  </forward>
  <bridge name='virbr1' stp='on' delay='0' />
  <mac address='52:54:00:BE:EE:16'/>
  <domain name='xxx.xxx.me'/>
  <ip address='xxx.xxx.xxx.xxx' prefix='28'>
    <tftp root='/var/lib/tftpboot' />
    <dhcp>
      <range start='xxx.xxx.xxx.242' end='xxx.xxx.xxx.254' />
      <bootp file='pxelinux.0' />
    </dhcp>
  </ip>
</network>


E, [2014-06-04T11:29:21.392750 #4056] ERROR -- : DHCP virsh provider error: unable to retrive virsh info: undefined method `<' for nil:NilClass
D, [2014-06-04T11:29:21.392801 #4056] DEBUG -- : /usr/share/foreman-proxy/bin/../lib/proxy/dhcp/server/virsh.rb:30:in `loadSubnets'
/usr/share/foreman-proxy/bin/../lib/proxy/dhcp/server.rb:32:in `load'
/usr/share/foreman-proxy/bin/../lib/proxy/dhcp/server.rb:36:in `subnets'
/usr/share/foreman-proxy/bin/../lib/dhcp_api.rb:24:in `dhcp_setup'
/usr/share/foreman-proxy/bin/../lib/dhcp_api.rb:38
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:461:in `instance_eval'
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:461:in `before_filter!'
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:461:in `each'
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:461:in `before_filter!'
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:600:in `dispatch!'
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:411:in `call!'
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:566:in `instance_eval'
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:566:in `invoke'
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:566:in `catch'
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:566:in `invoke'
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:411:in `call!'
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:399:in `call'
/usr/lib/ruby/gems/1.8/gems/rack-1.1.0/lib/rack/methodoverride.rb:24:in `call'
/usr/lib/ruby/gems/1.8/gems/rack-1.1.0/lib/rack/showexceptions.rb:24:in `call'
/usr/lib/ruby/gems/1.8/gems/rack-1.1.0/lib/rack/commonlogger.rb:18:in `call'
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:979:in `call'
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:1005:in `synchronize'
/usr/lib/ruby/gems/1.8/gems/sinatra-1.0/lib/sinatra/base.rb:979:in `call'
/usr/lib/ruby/gems/1.8/gems/rack-1.1.0/lib/rack/content_length.rb:13:in `call'
/usr/lib/ruby/gems/1.8/gems/rack-1.1.0/lib/rack/handler/webrick.rb:48:in `service'
/usr/lib/ruby/1.8/webrick/httpserver.rb:104:in `service'
/usr/lib/ruby/1.8/webrick/httpserver.rb:65:in `run'
/usr/lib/ruby/1.8/webrick/server.rb:173:in `start_thread'
/usr/lib/ruby/1.8/webrick/server.rb:162:in `start'
/usr/lib/ruby/1.8/webrick/server.rb:162:in `start_thread'
/usr/lib/ruby/1.8/webrick/server.rb:95:in `start'
/usr/lib/ruby/1.8/webrick/server.rb:92:in `each'
/usr/lib/ruby/1.8/webrick/server.rb:92:in `start'
/usr/lib/ruby/1.8/webrick/server.rb:23:in `start'
/usr/lib/ruby/1.8/webrick/server.rb:82:in `start'
/usr/lib/ruby/gems/1.8/gems/rack-1.1.0/lib/rack/handler/webrick.rb:14:in `run'
/usr/share/foreman-proxy/bin/../lib/sinatra-patch.rb:42:in `run!'
/usr/share/foreman-proxy/bin/smart-proxy:44
```

lzap commented:
Thanks for your fix, will test.

lzap commented:
Tested, works fine. [test]

We do carry monkey-patches to the IPAddr already in `lib/proxy/dhcp/monkey_patches.rb` - would you mind create a method there for this case? Might be handy in future and it is already full of ugly hacks :-)

thkrmr commented:
I'm going to take a look at monkey patching IPAddr.

thkrmr commented:
Monkey patch contains an alias method now, it shouldn't break any functionality in my opinion, since previous behavior wasn't modified.

lzap commented:
Many thanks, it is almost what I want :-) Although I also don't see any side effects, this kind of monkey patching is dangerous and I'd rather avoid it. Can you just create a simple method that returns the mask?

I have created a ticket, please rebase and squash everything into one commit with this message:

```
fixes #6085 - fixed virsh dhcp provider netmask interpretation
```

or something like that. Then we are ready to go.

ohadlevy commented:
@thkrmr why close? :)

thkrmr commented:
Sorry, this was unintentional.

ohadlevy commented on lib/proxy/dhcp/monkey_patches.rb:
```diff
@@ -1,6 +1,11 @@
+  def to_mask
+    return _to_string(@mask_addr)
+  end
+  
```
note that this code will only executed based on the condition at https://github.com/thkrmr/smart-proxy/blob/f70a52b9f070da37c522150e6d01caf764f19395/lib/proxy/dhcp/subnet.rb#L3

thkrmr replied:
Sorry I don't get what you are trying to tell me, I've been chasing a different bug related to loadSubnetData for the past few hours, I'm not up to speed.

ohadlevy replied:
if you look at the latest version of the proxy code, it has been broken
down to allow multiple plugins (e.g outside of the initial proxy code).
this change conflict with yours (most code wasn't really changed, just
moved a bit), but will not allow a clean merge, in git terms, the way to
fix it is called a rebase :)

ohadlevy commented:
thanks! I realize that since converting proxy into plugins, this now
requires a rebase :(

thkrmr committed:
fixes #6085 - fixed virsh dhcp provider netmask interpretation

The virsh.rb loadSubnets method was only looking for a definition of
a netmask inside the libvirt network XML, despite netmask not being
there in case a prefix attribute was defined. netmask and prefix are
mutually exclusive by libvirt definition.

Also the ipaddr library has no easy way of retrieving a readable netmask
from a prefix, albeit this functionality is provided by the inspect
method of IPAddr, a monkey-patch "to_mask" was introduced to circument
the original implementation.

thkrmr committed:
fixes #6377 - incorrect dhcp providers server virsh.rb loadSubnetData

The commit fixes dhcp server virsh.rb provider. Upon loading the
contents of the libvirt network XML and parsing for DHCP reservations a
check is performed on the kind of records that are being loaded. The
dhcp_api.rb 'delete' helper tries to load network entries of the
'reservation' kind, the helper uses modules/dhcp/dhcp/subnet.rb to
validate on that and filter out entries.

In contrary to other providers (ms, isc) the loadSubnetData method of
virsh.rb was dealing out "Proxy::DHCP::Record" for any entry it was able
to discover, but should have been instantiating
"Proxy::DHCP::Reservation" to pass the validation.

This resulted in "ERROR -- : Record
x.x.x.x/xx:xx:xx:xx:xx:xx not found" as well as leaves in the libvirt
network XML. This prevented the creation of VMs with the same addresses,
since leftover DHCP entries conflicted with the new ones.

lzap commented:
Looking good, I've created another ticket, if you can rebase the commit message and add the reference and we are ready to merge! http://projects.theforeman.org/issues/6377

lzap commented:
Thanks for this nice patch. [test]

domcleal commented:
[test] @lzap

lzap commented:
I had a week off, merged as 
- 1fb2109 fixes #6377 - incorrect dhcp providers server virsh.rb loadSubnetData <Vincent Kramar>
- 499c380 fixes #6085 - fixed virsh dhcp provider netmask interpretation <Vincent Kramar>

thanks for your contribution!
